ts/nni_manager/core/nnimanager.ts

```diff
@@ -296,10 +297,70 @@ class NNIManager implements Manager {
     }
 
     public async stopExperiment(): Promise<void> {
+        await this.stopExperimentTopHalf();
+        await this.stopExperimentBottomHalf();
+    }
+
+    public async stopExperimentTopHalf(): Promise<void> {
```

Comment:
The API sounds very internal. Can we think of a formal name?

Reply:
The name comes from Linux interrupt handler. Top-half means it should be completed before handling control to caller.